### ISSUE TITLE
Simultaneous the object detection (instance segmentation) and road detection (semantic segmentation)

### ISSUE THREAD

**Issue by alexbuyval** [dbolya/yolact#73]
Hi
@dbolya Thank you for your impressive work!

I would like to try your approach to detect simultaneously objects and road. What do you think which approach more suitable?

1. label the road as a regular object with a mask and BB then to train your network without any changes;
2. add additional head for semantic segmentation.

Maybe there is another approach?

Thank you in advance!
Regards,
Alex

**Comment by dbolya** [comment]
Hmm, good question. I think the second approach would be more reasonable since you are likely to get more than one road detection using the first approach (since NMS isn't perfect). However, you actually don't need to decide because we already have a hidden semantic segmentation branch! I haven't tested how good it is because we only use it for an extra loss signal during training, but you can take a look.

Here is the relevant line:
https://github.com/dbolya/yolact/blob/13b49d749b734b098a292c8c5226017b344ccc67/yolact.py#L602

If you want to use it, just comment out that check for `self.training` and pass it through `self.detect`

**Comment by alexbuyval** [comment]
@dbolya Thank you for the quick response! 
I will try it!

**Comment by abhigoku10** [comment]
@alexbuyval where you able to perform semantic segmentation with the current code base , if so can let me know what all changes have to be made

**Comment by abhigoku10** [comment]
@dbolya  can you please elaborate how to expose the semantic segmentation output since when i am doing eval with the changes made i am stil obtaining the instance segmentation only ? should i re-train the model with the obtained changes or will i be able to obtain theoutput with the currrent eval code

**Comment by dbolya** [comment]
@abhigoku10 The current eval code does not expose the semantic segmentation branch at all. Like I said before, you'll need to expose it yourself by propagating this line:
https://github.com/dbolya/yolact/blob/605b3c8fdf283642e7753c355bc6ee02f8db1380/yolact.py#L619
through all the functions to get to eval.

**Comment by abhigoku10** [comment]
@dbolya to just to summarise , you want me to add " pred_outs['segm'] = self.semantic_seg_conv(outs[0]) " line and run my eval network for the current base model which will give me the semantic output

**Comment by dbolya** [comment]
@abhigoku10 Huh? No I mean that's where the segmentations are: in `pred_outs['segm']`. You have to pass that through `detection.py` and through the `evaluate` function in `eval.py` (depending on where you want it to end up). The line is already there, you just have to pass its outputs through the code.

**Comment by abhigoku10** [comment]
@dbolya  i followed the instructions and coded till eval.py and i am in post processing section, how to visualize it ? should i normally re use the visualize done for instance seg output

**Comment by dbolya** [comment]
@abhigoku10 You'll have to either create a box and class label for it and treat it like an instance segmentation, or write your own visualization code. Naturally, the code I have for instance segmentation visualization won't work 100% with semantic segmentation. Though, I recommend using matplotlib to quickly display the semantic segmentations to make sure everything's working.

**Comment by abhigoku10** [comment]
@dbolya i have used matplotlib and displayed the output , but for a single input image in post processing i am getting tensors of 80x69x69 (classes X image dimension). how to obtain the semantic mask of it . Is there any way you can share the semantic segmented code which you have tested it would be helpful

**Comment by dbolya** [comment]
@abhigoku10 Those *are* the semantic masks. Just interpolate them up to the size of the full image (using `torch.nn.functional.interpolate`). I don't have any semantic segmentation testing code.

**Comment by abhigoku10** [comment]
@dbolya thanks for the pointers i shall do that and revert back to u

**Comment by abhigoku10** [comment]
@dbolya  i am getting the below output 
![Seg](https://user-images.githubusercontent.com/22094053/64320020-44c41780-cfdb-11e9-9539-f48eadad9e99.png)

as you can see the instance mask is very good but the semantic mask is not that accurate . is this the expected output??

**Comment by dbolya** [comment]
Other than the foot, they look pretty similar I think (you'd just have to overlay them to tell). (oh and of course the handle of the tennis racket). But yeah, the semantic segmenter gets a lot less focus during training so it would make sense that it's worse. You could try giving it more focus by increasing the semantic segmentation alpha (idr the exact name of the config setting off the top of my head). Right now it's just set to 1.

**Comment by abhigoku10** [comment]
so shall i increase the semantic segmentation alpha value >1 and re-train the model ? Moreover the dataset is trained on MSCOCO instance segmentation , does it seem logical to re-train the model with semantic segmentation dataset and check the output

**Comment by dbolya** [comment]
I don't think it matters whether you train it on a specifically semantic segmentation dataset (unless you want to compare it against other methods), since if you have an instance segmentation it's easy to construct a semantic segmentation from that.

And yeah, if you need this semantic segmentation output to be high quality, I'd recommend increasing that alpha value. Right now if you train the model you can see that the semantic segmentation loss immediately goes to something really low (I think ~0.4) around 1k iterations in, so it might not contribute much past that with the current alpha.
